### PR TITLE
IBX-154: Allowed to define if the alternative text for image field is required

### PR DESCRIPTION
**Pull request by adamwojs**
| Question                                  | Answer
| ---------------------------------------- | ------------------
| **JIRA issue**                          | https://issues.ibexa.co/browse/IBX-154
| **Type**                                   | feature
| **Target eZ Platform version** | `v4.0`
| **BC breaks**                          | no
| **Doc needed**                       | yes

#### Checklist:
- [ ] Provided PR description.
- [x] Tested the solution manually.
- [x] Provided automated test coverage.
- [x] Checked that target branch is set correctly (master for features, the oldest supported for bugs).
- [x] Ran PHP CS Fixer for new PHP code (use `$ composer fix-cs`).
- [x] Asked for a review (ping `@ezsystems/php-dev-team`).

### CODE REVIEW

**Comment by adamwojs** [comment]
CI issues seems to be not related to PR changes.

File: eZ/Publish/Core/Persistence/Legacy/Content/FieldValue/Converter/ImageConverter.php [hunk 1]
```diff
@@ -176,6 +179,32 @@ public function toFieldValue(StorageFieldValue $value, FieldValue $fieldValue)
...
+        $storageDef->dataInt2 = (int)($validators['AlternativeTextValidator']['required'] ?? 0);
+    }
+
+    public function toFieldDefinition(StorageFieldDefinition $storageDef, FieldDefinition $fieldDef): void
+    {
+        $fieldDef->fieldTypeConstraints = new FieldTypeConstraints(
+            [
+                'validators' => [
+                    'FileSizeValidator' => [
+                        'maxFileSize' => ($storageDef->dataInt1 != 0
```
Comment (ViniTou) [inline eZ/Publish/Core/Persistence/Legacy/Content/FieldValue/Converter/ImageConverter.php:196]: What about strong comparison here?
```suggestion
                        'maxFileSize' => ($storageDef->dataInt1 !== 0
```
Not sure if it's only me but those parentheses are adding this feeling that there is something more here, but they seems redundant?

File: eZ/Publish/Core/FieldType/Tests/ImageTest.php [hunk 2]
```diff
@@ -817,6 +823,56 @@ public function provideInvalidDataForValidate()
...
+                ),
+                [
+                    new ValidationError(
+                        'Alternative text is required.', null, [], 'alternativeText'
+                    ),
+                ],
+            ],
+
+            // alternative text is empty string
+            [
```
Comment (Steveb-p) [inline eZ/Publish/Core/FieldType/Tests/ImageTest.php:853]: I think it makes sense for these cases to have this comment as part of the key for the array. Even if the previous are not "keyed", it's still possible PHP-wise to add set keys explicitly here, so previous ones do not need to be updated if you don't want to.

```suggestion
            'alternative text is empty string' => [
```

File: eZ/Publish/Core/Persistence/Legacy/Content/FieldValue/Converter/ImageConverter.php [hunk 3]
```diff
@@ -176,6 +179,32 @@ public function toFieldValue(StorageFieldValue $value, FieldValue $fieldValue)
...
 
+    public function toStorageFieldDefinition(FieldDefinition $fieldDef, StorageFieldDefinition $storageDef): void
+    {
+        $validators = $fieldDef->fieldTypeConstraints->validators;
+
+        $storageDef->dataInt1 = $validators['FileSizeValidator']['maxFileSize'] ?? 0;
+        $storageDef->dataInt2 = (int)($validators['AlternativeTextValidator']['required'] ?? 0);
+    }
+
+    public function toFieldDefinition(StorageFieldDefinition $storageDef, FieldDefinition $fieldDef): void
```
Comment (Steveb-p) [inline eZ/Publish/Core/Persistence/Legacy/Content/FieldValue/Converter/ImageConverter.php:190]: As a general remark, those "to*" methods would probably be better if they could return the `FieldDefinition` and `StorageFieldDefinition` respectively. I'm not sure what context they are used in, but I think usually those objects are constructed just before this method and are mutated here?

I.e.
```php
    public function toFieldDefinition(StorageFieldDefinition $storageDef, ?FieldDefintion $fieldDef = null): FieldDefinition
    {
        $fieldDef = $fieldDef ?? new FieldDefinition();
        // ...
        return $fieldDef;
    }
```

I understand that this is part of some sort of an interface, so only wanted to comment on it, nothing more.

**Review by Steveb-p — APPROVED** [review]

File: eZ/Publish/Core/Persistence/Legacy/Content/FieldValue/Converter/ImageConverter.php (hunk 1 above)
Comment (Steveb-p) [reply eZ/Publish/Core/Persistence/Legacy/Content/FieldValue/Converter/ImageConverter.php:196]: If that was supposed to be a "fix" for a float comparison then `abs($storageDef->dataInt1) < PHP_FLOAT_EPSILON` should be used instead?

Comment (ViniTou) [reply eZ/Publish/Core/Persistence/Legacy/Content/FieldValue/Converter/ImageConverter.php:196]: I am almost sure that `maxFileSize` is int not float.
And you cant store floats in `dataIntX` properties, right? We have `dataFloatX` for that.

Comment (alongosz) [reply eZ/Publish/Core/Persistence/Legacy/Content/FieldValue/Converter/ImageConverter.php:196]: > I am almost sure that `maxFileSize` is int not float.
> And you cant store floats in `dataIntX` properties, right? We have `dataFloatX` for that.

`dataIntX` should never contain anything other than int or null.

File: eZ/Publish/Core/Persistence/Tests/Legacy/FieldValue/Converter/ImageConverterTest.php [hunk 4]
```diff
@@ -225,4 +228,142 @@ public function xmlToFieldValueProvider(): array
...
+     */
+    public function testToStorageFieldDefinition(
+        FieldDefinition $fieldDefinition,
+        StorageFieldDefinition $expectedStorageDef
+    ): void {
+        $storageFieldDefinition = new StorageFieldDefinition();
+
+        $this->imageConverter->toStorageFieldDefinition($fieldDefinition, $storageFieldDefinition);
+
+        $this->assertEquals(
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Tests/Legacy/FieldValue/Converter/ImageConverterTest.php:243]:
```suggestion
        self::assertEquals(
```

File: eZ/Publish/Core/Persistence/Tests/Legacy/FieldValue/Converter/ImageConverterTest.php [hunk 5]
```diff
@@ -225,4 +228,142 @@ public function xmlToFieldValueProvider(): array
...
+     */
+    public function testToFieldDefinition(
+        StorageFieldDefinition $storageDef,
+        FieldDefinition $expectedFieldDefinition
+    ): void {
+        $fieldDefinition = new FieldDefinition();
+
+        $this->imageConverter->toFieldDefinition($storageDef, $fieldDefinition);
+
+        $this->assertEquals(
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Tests/Legacy/FieldValue/Converter/ImageConverterTest.php:323]:
```suggestion
        self::assertEquals(
```

**Review by alongosz — CHANGES_REQUESTED** [review]
Please format PR title & commit according to the existing convention

**Comment by adamwojs** [comment]
PR updated according to code review suggestions.

**Review by ViniTou — APPROVED** [review]

**Review by alongosz — APPROVED** [review]

**Comment by adamwojs** [comment]
Changes integrated manually.

**Pull request closed without merging**